### PR TITLE
added typing to linear_algebra.py

### PR DESCRIPTION
**Pull request by shivance**
I added types to `linear_algebra.py`

I'm not sure whether we should add types to nested private functions, so I haven't for now.
Please review !

(#250 )

### CODE REVIEW

File: optax/_src/linear_algebra.py
Comment (mkunesch) [inline optax/_src/linear_algebra.py:36]: `num_iters` has a default value but I don't think it has `Optional` type. In particular, `Optional[T]` is just `Union[None, T]` and I think passing None to `num_iters` would lead to an error in line 61: `jnp.logical_and(i < num_iters, run_step)`.

File: optax/_src/linear_algebra.py
Comment (mkunesch) [inline optax/_src/linear_algebra.py:37]: As above and for the other types you have marked as Optional too - please check whether it would be possible/intended to pass `None`.

File: optax/_src/linear_algebra.py
Comment (mkunesch) [inline optax/_src/linear_algebra.py:91]: Afaik `lax.Precision` is an Enum. That means we could very nicely enforce the 3 available options (DEFAULT, HIGH, HIGHEST) by typing this as `lax.Precision`.

**Review by mkunesch — COMMENTED** [review]
Thanks a lot for the PR!

**Review by mkunesch — CHANGES_REQUESTED** [review]
(sorry, forgot to tick the box to mark this as reviewed)

**Comment by shivance** [comment]
Hey @mkunesch  ! Thanks for the feedback . Got the point over there wrt Optional !
Performed the requested change.

**Comment by shivance** [comment]
@mkunesch  Thanks for the feedback . Got the point over there wrt Optional !
Performed the requested change.

File: optax/_src/linear_algebra.py
Comment (hbq1) [inline optax/_src/linear_algebra.py:22]: Could you remove this line to make CI tests pass?

**Comment by shivance** [comment]
@hbq1 I have removed the unused import ! and the PR is ready for next round of review once the CI passes.

**Review by hbq1 — APPROVED** [review]
Thanks!

**Comment by shivance** [comment]
@hbq1 the CI turned out to be clean .

**Comment by hbq1** [comment]
Nice! Waiting for approval from @mkunesch

**Pull request merged**
